tests/python/pants_test/cache/test_pinger.py

```diff
@@ -4,68 +4,96 @@
+    latency = cls.latency_by_url[url]
+
+    # TODO(John Sirois): Switch to a homomorphic response in each branch once
+    #   https://github.com/getsentry/responses/issues/234 is fixed.
+    response = requests.exceptions.ConnectTimeout() if latency >= timeout else (200, {}, '')
+
+    def callback(_):
+      if latency < timeout:
+        times.append(latency)
+      else:
+        # We raise a timeout exception.
+        pass
+      return response
+
+    responses.add_callback(responses.HEAD, url, callback)
 
+  @contextmanager
+  def pinger(self, timeout, urls, tries=2):
+    with mock.patch('pants.cache.pinger.Timer.elapsed', new_callable=mock.PropertyMock) as elapsed:
+      times = []
+      for url in urls:
+        for _ in range(tries):
+          self.expect_response(url, timeout, times)
+
+      elpased_results = iter(times)
+      elapsed.side_effect = elpased_results
+
+      yield Pinger(timeout=timeout, tries=tries)
+
+      # Ensure our mock Timer was used exactly the number of times we expected.
+      with self.assertRaises(StopIteration):
+        next(elpased_results)
+
+  @responses.activate
+  def test_pinger_times_correct(self):
+    urls = [self.fast_url, self.slow_url, self.unreachable_url]
+    with self.pinger(timeout=0.4, urls=urls) as test:
+      ping_results = dict(test.pings(urls))
+      self.assertEqual(ping_results[self.slow_url], 0.3)
+      self.assertLess(ping_results[self.fast_url], ping_results[self.slow_url])
+      self.assertEqual(ping_results[self.unreachable_url], Pinger.UNREACHABLE)
+
+  @responses.activate
   def test_pinger_timeout_config(self):
-    test = Pinger(timeout=self.fast_timeout_seconds, tries=2)
-    netlocs = [self.fast_netloc, self.slow_netloc]
-    ping_results = dict(test.pings(netlocs))
-    self.assertLess(ping_results[self.fast_netloc], self.fast_timeout_seconds)
-    self.assertEqual(
-      ping_results[self.slow_netloc], Pinger.UNREACHABLE, msg=self.message)
+    urls = [self.fast_url, self.slow_url]
+    with self.pinger(timeout=0.2, urls=urls) as test:
+      ping_results = dict(test.pings(urls))
+      self.assertEqual(ping_results[self.fast_url], 0.1)
+      self.assertEqual(ping_results[self.slow_url], Pinger.UNREACHABLE)
 
+  @responses.activate
   def test_global_pinger_memo(self):
-    fast_pinger = Pinger(timeout=self.fast_timeout_seconds, tries=2)
-    slow_pinger = Pinger(timeout=self.slow_timeout_seconds, tries=2)
-    self.assertEqual(
-      fast_pinger.pings([self.slow_netloc])[0][1], Pinger.UNREACHABLE, msg=self.message)
-    self.assertNotEqual(
-      slow_pinger.pings([self.slow_netloc])[0][1], Pinger.UNREACHABLE, msg=self.message)
+    urls = [self.slow_url]
+    with self.pinger(timeout=0.2, urls=urls) as fast_pinger:
+      self.assertEqual(fast_pinger.pings([self.slow_url])[0][1], Pinger.UNREACHABLE)
+    with self.pinger(timeout=0.4, urls=urls) as slow_pinger:
+      self.assertNotEqual(slow_pinger.pings([self.slow_url])[0][1], Pinger.UNREACHABLE)
```

Comment:
Is there a meaningful way to rewrite this line to avoid a negative assertion? What we discussed in Slack https://nedbatchelder.com/blog/201811/careful_with_negative_assertions.html

I'm conflicted in requesting this, because I do like how this reads now. There's a very clear distinction between `fast_pinger` and `slow_pinger` that makes it explicit what the test is doing. I think this may be a case that the negative assertion is valid, only wanted to bring it up if you or anyone has alternative ideas.

Reply:
Yes - thanks. Using `assertLessThan` proves the same point better - fixed.